### PR TITLE
Add additional fields to GCPManagedMachinePool

### PR DESCRIPTION
**Pull request by akshay196-rafay**
**What type of PR is this?**

/kind api-change

**What this PR does / why we need it**:

Support additional fields to GCPManagedMachinePool kind.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #906

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [X] includes documentation

**Release note**:

```release-note
Support additional fields to GCPManagedMachinePool type.
```

### CODE REVIEW

**Comment by richardcase** [comment]
@akshay196-rafay - i think we should perhaps re-think the approach going forward. So that we split the aditions up into smaller PRs where each contains:

- A small number of API additions
- Reconciliation changes for the new API fields
- Updates to the e2e

I appreciate this is following the original approach for the GKE implementation but  i think it may be better for to do this incrementally.

wdyt?

**Comment by akshay196-rafay** [comment]
> i think we should perhaps re-think the approach going forward. So that we split the aditions up into smaller PRs where each contains:
>
>    A small number of API additions
>    Reconciliation changes for the new API fields
>    Updates to the e2e
>
> I appreciate this is following the original approach for the GKE implementation but i think it may be better for to do this incrementally.

I understand where you're coming from. Since GCPManagedMachinePool's new fields are comparatively less I can continue working on this same PR to add reconciliation changes, validation webhooks and e2e of GCPManagedMachinePool.

For the GCPManagedControlPlane [PR](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/936) raised by @aniruddha2000  we can take the incremental approach something like this - 

**Steps 1**
Add new API fields - Description, ClusterNetwork, DefaultNodeLocation, DefaultMaxPodsPerNode
Add reconciliation and webhook for same
Update e2e for these fields

**Steps 2**
Add new API fields - ClusterSecurity, AddonsConfig, LoggingConfig, MonitoringConfig
Add reconciliation and webhook for same
Update e2e for these fields

**Comment by richardcase** [comment]
@akshay196-rafay - sounds good to me 👍

File: cloud/services/container/nodepools/reconcile.go
Comment (richardcase) [inline cloud/services/container/nodepools/reconcile.go:336]: It would be good to warp this error to give some extra context.

File: exp/api/v1beta1/gcpmanagedmachinepool_types.go
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedmachinepool_types.go:53]: When we have a enum like this we normally create a new type (from a string in this case) and declare constants for each of the possible values. Like this: 

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/2e5725a57078e0778939e7105c4303b5607a7377/exp/api/v1beta1/gcpmanagedcontrolplane_types.go#L110-L120

File: exp/api/v1beta1/gcpmanagedmachinepool_types.go
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedmachinepool_types.go:126]: If you are specifying the service account config in `NodeSecurityConfig` then do you have to supply one of the fields in this struct?

File: exp/api/v1beta1/gcpmanagedmachinepool_types.go
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedmachinepool_types.go:57]: Should we have a min value validation on this as well to stop people entering a negative or zero value?

File: exp/api/v1beta1/gcpmanagedmachinepool_types.go
Comment (richardcase) [inline exp/api/v1beta1/gcpmanagedmachinepool_types.go:61]: Do we need a min value validation here as well?

File: cloud/services/container/nodepools/reconcile.go
Comment (gzcharleszhang) [inline cloud/services/container/nodepools/reconcile.go:169]: The `checkDiffAndPrepareUpdateVersionOrImage` block actually calls `UpdateNodePool` in the underlying logic. I think we can combine these so we're updating fields in one API call. Otherwise the second update call will block on the first. I'm doing something like this in my PR https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/1026/files#diff-1d279184a2f22d9c46b9f844a1852f6a04956ef24533b4f066bc9e900abd71bdR147.

File: cloud/services/container/nodepools/reconcile.go
Comment (akshay196) [reply cloud/services/container/nodepools/reconcile.go:314]: Updated `updateNodePool` caller function. Hope it is fine now.

File: exp/api/v1beta1/gcpmanagedmachinepool_types.go
Comment (akshay196) [reply exp/api/v1beta1/gcpmanagedmachinepool_types.go:141]: You can either specify `Email` which is Google Cloud Platform Service Account to be used, or specify `Scopes` for default service account (email of GKE default account or email field left blank).

**Comment by akshay196-rafay** [comment]
@richardcase @gzcharleszhang Addressed your review comments.

**Review by gzcharleszhang — COMMENTED** [review]
1 nit comment otherwise LGTM. I'll let @richardcase and @cpanato take a look.

File: cloud/services/container/nodepools/reconcile.go
Comment (akshay196-rafay) [reply cloud/services/container/nodepools/reconcile.go]: Done.

File: cloud/services/container/nodepools/reconcile.go
Comment (richardcase) [inline cloud/services/container/nodepools/reconcile.go:359]: Can we change this to use cmp.Equals?

(and also the other instances below)

**Review by richardcase — COMMENTED** [review]
Thanks @akshay196-rafay , this looks good to me. Just one small nit.

File: cloud/services/container/nodepools/reconcile.go
Comment (akshay196-rafay) [reply cloud/services/container/nodepools/reconcile.go]: Updated.

**Comment by richardcase** [comment]
Thanks @akshay196-rafay. From my side:

/approve

What do you think @gzcharleszhang or @richardchen331 ?

**Comment by richardcase** [comment]
@cpanato - in case you have some spare cycles to have a look and lgtm if you think its ok.

**Review by gzcharleszhang — APPROVED** [review]
LGTM

**Comment by richardchen331** [comment]
> Thanks @akshay196-rafay. From my side:
> 
> /approve
> 
> What do you think @gzcharleszhang or @richardchen331 ?

LGTM. Thanks for the contribution @akshay196-rafay!

**Review by richardchen331 — APPROVED** [review]

**Review by cpanato — APPROVED** [review]
thanks so much and sorry for the delay

/lgtm

**Comment by akshay196-rafay** [comment]
Thank you @richardcase, @gzcharleszhang, @richardchen331 and @cpanato for reviewing and merging this.

**Pull request merged**
